sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void foo() {
+    double d = 3.14;
+  }
+
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) void bar() {}
+};
+
+sycl::queue q(/* device selector returns a device *without* fp64 support */);
+assert(!q.get_device().has(sycl::aspect::fp64));
+
+q.single_task(syclext::properties{syclext::calls_indirectly<set_A>}, [=] {
+  // Exception is expected to be thrown, because target device doesn't support
+  // fp64 aspect and it is required by 'Foo::foo' which is included into 'set_A'
+});
+
+q.single_task(syclext::properties{syclext::calls_indirectly<set_B>}, [=] {
+  // No exceptions are expected, because 'set_B' doesn't bring any requirements
+  // for optional kernel features.
+});
+----
+
+An implementation may not raise a compile time diagnostic or a run time
+exception merely due to speculative compilation of a virtual member function for
+a device when the application does not specify a use of virtual member functions
+through the correponding properties.
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct Foo {
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {
+  double d = 3.14;
+}
+
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void bar() {}
+};
+
+sycl::queue q(/* device selector choosing a device *without* fp64 support */);
+assert(!q.get_device().has(sycl::aspect::fp64));
+
+auto *Storage = sycl::malloc_device<Foo>(1, q);
+
+q.single_task([=] {
+  // The kernel is not submitted with 'calls_indirectly' property and therefore
+  // it is not considered to be using any of virtual member functions of 'Foo'.
+  // This means that the object of 'Foo' can be successfully created by this
+  // kernel, regardless of whether a target device supports 'fp64' aspect which
+  // is used by 'Foo::foo'.
+  // No exceptions are expected to be thrown.
+  new (Storage) Foo;
+});
+----
+
+==== Interaction with `reqd_sub_group_size` attribute
+
+The `reqd_sub_group_size` attribute is a bit of a special case comparing to
+other optional kernel features, because it requires to compile a kernel in a
+certain way, which may require special handling for all functions which are
+called from it.
+
+When the same helper function is called from two or more kernels
+with different `reqd_sub_group_size` attribute, it may be required for the
+implementation to duplicate that helper function to create different versions
+of it tailored to different sub-group sizes. It can be done in a straightforward
+manner when operating on a static call graph.
+
+Virtual member functions are essentially called indirectly and pointers to them
+are initialized just once when an object of a polymorhpic class is being
+created. Therefore, to support calling such virtual member function from two or
+more kernels with different `reqd_sub_group_size`, each kernel may need to
+receive a different pointer to a different version of a virtual member function.
```

Comment:
I'm not sure that I'm reading this correctly. A polymorphic object contains a single pointer to the virtual function table associated with the dynamic type of the class (in the case of multiple inheritance, there may be more pointers, but that doesn't matter here). The virtual function table is static and built at compile time. When a polymorphic object is constructed, its (single) virtual function table pointer is updated as construction begins for each base class progressing through the most derived class. So it isn't correct to say that "pointers to them
are initialized just once when an object of a polymorphic class is being created." I don't see what `reqd_sub_group_size` has to do with this.

Reply:
`reqd_sub_group_size` attribute is effectively lowered into vectorization width for a kernel. The main question here is what happens when we have several kernels with different `reqd_sub_group_size` attribute calling the same virtual function? There has to be some sort of contract within the compiler so that all kernels (which are vectorized differently) can perform call to the same function (which may be vectorized in some other way distinct from all kernels).

Moreover, when a function is called from a kernel with certain `reqd_sub_group_size`, then it is expected that all sub-group built-ins will properly work within it, like querying sub-group size should return the right value regardless of which kernel the function was called from.